### PR TITLE
PLT-8171: Combine PIR, PLC, and UPLC into a single executable

### PR DESCRIPTION
**Pull request by bezirg**
DON'T LOOK HERE YET

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Changelog fragments have been written (if appropriate)
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting master unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

File: plutus-core/executables/plutus/AnyProgram/Compile.hs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    -- self to self patterns
+    (SPir SName a1, SPir SName a2) ->
+        through pirTypecheck
+        >=> toOutAnn a1 a2
+    (SPlc n1 a1, SPlc n2 a2) ->
+        through plcTypecheck
+        >=> plcToOutName n1 n2
+        >=> toOutAnn a1 a2
+    (SUplc n1 a1, SUplc n2 a2) ->
+        through (modifyError (fmap PIR.Original) . uplcTypecheck n1 a1)
```
Comment (zliu41) [inline plutus-core/executables/plutus/AnyProgram/Compile.hs:75]: nit: for UPLC it's not typecheck, but checking for unique errors.

File: plutus-core/executables/plutus/AnyProgram/Compile.hs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+    -- pir to plc
+    (sng1@(SPir n1@SName a1), SPlc n2 a2@SUnit) ->
+        -- prove that a1 has Ord,Pretty
+        withOrdPrettyA a1 $
+                (PLC.runQuoteT . flip runReaderT (PIR.toDefaultCompilationCtx undefined) . PIR.compileProgram)
+            -- get rid of annotations for now
+            >=> pure . void
+            >=> plcToOutName n1 n2
+    (SPir SName _, SPlc _ _) -> const $ throwing PIR._Error $ PIR.OptionsError "only support unit-ann output for now"
```
Comment (zliu41) [inline plutus-core/executables/plutus/AnyProgram/Compile.hs:48]: It should be straightforward to allow output ann type to be the same as the input ann type, isn't it?

File: plutus-core/executables/plutus/GetOpt.hs [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+import PlutusPrelude
+import Control.Monad
+import System.Console.GetOpt as GetOpt
+import Control.Lens
+import Data.Monoid
+
+data Opts = Opts
+    { _inputs :: [SomeFile]
+    , _output :: Maybe SomeFile
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/GetOpt.hs:27]: I added comments:

-- | This represents the parsed options retrieved from `GetOpt`.
--
-- After getopt parsing is done `Opts` may still contain erroneous options
-- (e.g. `_output=Nothing`). It is the responsibility of each mode to later cleanup or
-- fail on such erroneous `Opts` value.
data Opts = Opts
    { _inputs :: [SomeFile]
    , _output :: Maybe SomeFile -- ^ Nothing means the user didn't provide any output

File: plutus-core/executables/plutus/GetOpt.hs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    , parseArgs
+    , optDescrs
+    , GetOpt.usageInfo
+    ) where
+
+import Types
+
+import PlutusPrelude
+import Control.Monad
+import System.Console.GetOpt as GetOpt
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/GetOpt.hs:20]: I don't think the positional/stateful cmdargs parsing of -x / -a / -n would be possible with optparse-applicative, because of "applicativeness"

File: plutus-core/changelog.d/20240116_163710_bezirg_exe_combined.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+### Added
+
+- An experimental "plutus" tool that combines PIR,PLC, and UPLC compiling&manipulation into one executable file
```
Comment (zliu41) [inline plutus-core/changelog.d/20240116_163710_bezirg_exe_combined.md:3]:
```suggestion
- An experimental "plutus" tool that combines PIR, PLC, and UPLC compiling, manipulation and evaluation into one executable file
```

**Review by zliu41 — COMMENTED** [review]
Looks reasonable overall, except for `undefined` and the compile errors.

File: plutus-core/plutus-core/src/PlutusCore/Compiler/Types.hs [hunk 6]
```diff
@@ -15,5 +15,4 @@ type Compiling m uni fun name a =
   , Ord name
   , Typeable name
   , Hashable fun
-  , Hashable a
```
Comment (effectfully) [inline plutus-core/plutus-core/src/PlutusCore/Compiler/Types.hs:18]: Great spot.

File: plutus-core/plutus-core/src/PlutusCore/Compiler/Erase.hs [hunk 7]
```diff
@@ -1,10 +1,13 @@
 module PlutusCore.Compiler.Erase (eraseTerm, eraseProgram) where
 
 import PlutusCore.Core
+import PlutusCore.Name
 import UntypedPlutusCore.Core qualified as UPLC
 
 -- | Erase a Typed Plutus Core term to its untyped counterpart.
-eraseTerm :: Term tyname name uni fun ann -> UPLC.Term name uni fun ann
+eraseTerm :: HasUnique name TermUnique
```
Comment (effectfully) [inline plutus-core/plutus-core/src/PlutusCore/Compiler/Erase.hs:8]: Add a short comment please. And yeah, I think long-term we should fix debruijnification and undo this change. It's fine for now, though.

File: plutus-core/executables/plutus/Mode/PrintBuiltins.hs [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+-- following is PORTED from the older executable
+
+portedImpl :: IO ()
+portedImpl = do
+    -- MAYBE: categorize the builtins by Plutus Version introduced. Would require dependency
+    -- upon plutus-ledger-api
+    let builtins = enumerate @PLC.DefaultFun
+    mapM_
+      (\x -> putStr (printf "%-35s: %s\n" (show $ PP.pretty x) (show $ getSignature x)))
+      builtins
```
Comment (effectfully) [inline plutus-core/executables/plutus/Mode/PrintBuiltins.hs:32]: Nit: I'd write it as `for_ builtins $ \x -> <...>`.

File: plutus-core/executables/plutus/Mode/PrintBuiltins.hs [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+import Prettyprinter qualified as PP
+import Text.Printf
+import Data.Proxy (Proxy (..))
+import Data.List (intercalate)
+import GHC.TypeLits (symbolVal)
+
+runPrintBuiltins :: IO ()
+runPrintBuiltins = portedImpl
+
+-- following is PORTED from the older executable
```
Comment (effectfully) [inline plutus-core/executables/plutus/Mode/PrintBuiltins.hs:23]: Do we need this comment for something?

**Review by effectfully — APPROVED** [review]
Didn't review some of the details.

Looks OK to me, you could reduce the amount of boilerplate, but it doesn't really matter.

**Review by kwxm — COMMENTED** [review]
I don't fully understand this because of the singletons etc.  I have a few general comments though.

* It looks as if a lot of the complexity comes from the fact that `apply` takes multiple arguments.  Maybe we could just get rid of it and then (I think) every "command" would have at most one input and one output: that might let us go back to applicative parsing).  If we still want the `apply`  functionality it might be easier to have a separate executable for it.  I did need this in the past when I had unapplied validators and I wanted to supply them with `data` arguments, but I haven't used it recently.

* I like the idea of the smart suffixes, but they're almost not smart enough because we sometimes have to specify the name type and the annotation type.  It would be unbearable to have to have files called `fib.uplc.flat.NamedDeBruijn.unitAnnotations` though.  I had an idea of having some default format/name/annotation combination for each language and then letting the user override it with a sort of fake filename that says what those could be, like you have a file called `fib.flat` but the parameter you pass to `plutus` is something like `fib.flat@NamedDeBruijn@unitAnnotations`. Maybe that's no better than `-x`and friends though.

* I sometimes have a file and find myself trying to work out what's in it by running `uplc` with different combinations of options, which is pretty annoying.  The best way to deal with this would probably have been to have some wrapper round the file containing metadata that says what's in it, but it's probably too late to do that.   Maybe we could write a program that tries to read the input using different combinations until it succeeds, or even have a `--auto` option for `plutus` that does that for you.  I'm not sure if that would work though:  some failures may be untrappable.  

* Something else that I've forgotten!

File: plutus-core/executables/plutus/GetOpt.hs [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+    deriving stock Show
+makeLenses ''Opts
+
+parseArgs :: [String] -> IO Opts
+parseArgs args = do
+    let (getOptRes, _, getOptErrMsgs) = GetOpt.getOpt (GetOpt.ReturnInOrder fromNonDash) optDescrs args
+
+    when (not $ null getOptErrMsgs) $ failE $ fold getOptErrMsgs
+
+    -- MAYBE: I could make --stdout completely implicit when getOptRes.output==Nothing
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/GetOpt.hs:51]: I think it is possible to also do this if there is no output specified:

-  If there is no unix pipe connected after this command, output to a file with a specific name (a.out or a combination of all applied program names)
- if there is a pipe connected, do not write to any file like a.out  but just redirect the output to the pipe.

I do not know yet how to implement this, but I think it is possible.

File: plutus-core/executables/plutus/AnyProgram/With.hs [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+withShowBinderN s r = case s of
+    SName -> r
+    SDeBruijn -> r
+    SNamedDeBruijn -> r
+
+withShowN :: forall s r. SNaming s ->  (Show (FromName s) => r) -> r
+withShowN s r = case s of
+    SName -> r
+    SDeBruijn -> r
+    SNamedDeBruijn -> r
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/With.hs:29]: This is cool, but I tried it and failed. Do u have a concrete example that works?

This typechecks:

```
withN
    :: forall constr s r.
        (constr (FromName 'Name), constr (FromName 'DeBruijn), constr (FromName 'NamedDeBruijn))
    => SNaming s -> ((constr (FromName s)) => r) -> r
withN s r = case s of
    SName -> r
    SDeBruijn -> r
    SNamedDeBruijn -> r
```

but when I try to use it I get:

```
withShowN :: forall s r. SNaming s ->  (Show (FromName s) => r) -> r
withShowN s r = withN s r


executables/plutus/AnyProgram/With.hs:28:17: error: [GHC-05617]
    • Could not solve: (constr0 UPLC.Name, constr0 UPLC.DeBruijn,
                        constr0 UPLC.NamedDeBruijn)
        arising from a use of ‘withN’
    • In the expression: withN s r
      In an equation for ‘withShowN’: withShowN s r = withN s r
   |
28 | withShowN s r = withN s r
   |                 ^^^^^

executables/plutus/AnyProgram/With.hs:28:25: error: [GHC-39999]
    • Could not deduce ‘Show (FromName s)’ arising from a use of ‘r’
      from the context: constr0 (FromName s)
        bound by a type expected by the context:
                   constr0 (FromName s) => r
        at executables/plutus/AnyProgram/With.hs:28:25
    • In the second argument of ‘withN’, namely ‘r’
      In the expression: withN s r
      In an equation for ‘withShowN’: withShowN s r = withN s r
   |
28 | withShowN s r = withN s r
   |                         ^
Error: cabal: Failed to build exe:plutus from plutus-core-1.20.0.0.


```

[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/executables/plutus/AnyProgram/With.hs:29]: Yes, you have to specify the constraint explicitly, GHC cannot figure out which one you want (even though you specified it). Something like this should work:

```haskell
withShowN :: forall s r. SNaming s ->  (Show (FromName s) => r) -> r
withShowN s r = withN @Show s r
```

That's why `constr` is the first argument of `withN`.

[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/With.hs:29]: Awesome, will try it out

[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/With.hs:29]: OMG  it works

File: plutus-core/executables/plutus/AnyProgram/Example.hs [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+import PlutusCore.Default
+import PlutusCore.StdLib.Data.Bool qualified as StdLib
+import PlutusCore.StdLib.Data.ChurchNat qualified as StdLib
+import PlutusCore.StdLib.Data.Integer qualified as StdLib
+import PlutusCore.StdLib.Data.Unit qualified as StdLib
+
+-- TODO: generalize annotation after removal of Provenance
+-- TODO: had to constrain it to Name&TyName, use sth like plcViaName to overcome this
+termExamples :: [( ExampleName
+                , forall term. (TermLike term PLC.TyName PLC.Name DefaultUni DefaultFun) => term ()
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/Example.hs:20]: I was planning to instantiate based on the singleton (like if the user wanted to be pir,plc,uplc), but I don't have it working yet.

[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/Example.hs:20]: What buys me when I put it inside, is that I don't have to instantiate term at ListExample.hs  together with ImpredicativeTypes. Do you know another why?

File: plutus-core/executables/plutus/AnyProgram/With.hs (hunk 11 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/executables/plutus/AnyProgram/With.hs:29]: Sorry, what is it that you want to do and what error are you getting?

[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/With.hs:29]: I am trying to have a similar function like `withShowL`, etc
, but the constraint is `TermLike`.

I would like to prove that having a singleton SLang s , I can use the methods of MkPlc.TermLike on it.

File: plutus-core/executables/plutus/AnyProgram/Example.hs (hunk 12 above)
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/executables/plutus/AnyProgram/Example.hs:20]: > I was planning to instantiate this based on the input singleton (like if the user wanted to be pir,plc,uplc), but I don't have it working yet.

You can put the constraint outside then:

```haskell
termExamples :: forall term. (TermLike term PLC.TyName PLC.Name DefaultUni DefaultFun) => [( ExampleName, term ())]
```

> What buys me when I put it inside, is that I don't have to instantiate term at `ListExample.hs` together with `ImpredicativeTypes`.

Sorry, I don't understand.

File: plutus-core/executables/plutus/AnyProgram/With.hs (hunk 11 above)
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/With.hs:29]: If I had sth like `withTermLikeL` then I could print/compile examples based on user's choice (plc,pir,uplc)

File: plutus-core/executables/plutus/AnyProgram/Parse.hs [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+import UntypedPlutusCore qualified as UPLC
+
+import Data.Text qualified as T
+import Data.Singletons.Decide
+import Control.Monad.Except
+
+-- | Given a singleton witness and two programs of that type witness, apply them together.
+--
+-- This could alternatively be achieved by
+-- using a "Parsable" typeclass + withParsable hasomorphism
```
[the remark this answers is not in the document]
Comment (effectfully) [reply plutus-core/executables/plutus/AnyProgram/Parse.hs:26]: Must be a reference to the [Hasochism](https://homepages.inf.ed.ac.uk/slindley/papers/hasochism.pdf) paper

File: plutus-core/executables/plutus/AnyProgram/Example.hs (hunk 12 above)
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/Example.hs:20]: If I give your signature, ListExamples.hs complains with:

executables/plutus/Mode/ListExamples.hs:14:18: error: [GHC-39999]
    • Ambiguous type variable ‘term0’ arising from a use of ‘termExamples’
      prevents the constraint ‘(PlutusCore.MkPlc.TermLike
                                  term0
                                  PlutusCore.Name.TyName
                                  PlutusCore.Name.Name
                                  plutus-core-1.20.0.0:PlutusCore.Default.Universe.DefaultUni
                                  PlutusCore.Default.Builtins.DefaultFun)’ from being solved.
      Probable fix: use a type annotation to specify what ‘term0’ should be.

File: plutus-core/executables/plutus/Mode/PrintBuiltins.hs [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+
+portedImpl :: IO ()
+portedImpl = do
+    -- MAYBE: categorize the builtins by Plutus Version introduced. Would require dependency
+    -- upon plutus-ledger-api
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/Mode/PrintBuiltins.hs:28]: I don't think  a user cares about the BuiltinSemanticsVariant.

File: plutus-core/executables/plutus/AnyProgram/Compile.hs (hunk 1 above)
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/Compile.hs]: Y, i just kept it for "uniformity"

File: plutus-core/executables/plutus/AnyProgram/Apply.hs [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+--
+applyProgram :: (?opts :: Opts, MonadError PLC.ApplyProgramError m)
+             => SLang s -> FromLang s -> FromLang s -> m (FromLang s)
+applyProgram _ _ _ | _wholeOpt ?opts = error "error FIXME: not implemented yet"
+applyProgram sng p1 p2 = withSemigroupA (_sann sng) $
+    case sng of
+        SPir{} -> PIR.applyProgram p1 p2
+        SPlc{} -> PLC.applyProgram p1 p2
+        SUplc{} -> UPLC.UnrestrictedProgram <$> UPLC.applyProgram (UPLC.unUnrestrictedProgram p1) (UPLC.unUnrestrictedProgram p2)
+        SData{} -> error "FIXME: not implemented yet"
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/Apply.hs]: Yes, there is plan to move away from error and fail and go completely to pure errors. But i leave it for another PR

File: plutus-core/executables/plutus/AnyProgram/Compile.hs [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+plcToOutName :: (PLC.AsFreeVariableError e, MonadError e m)
+             => SNaming s1
+             -> SNaming s2
+             -> PLC.Program (FromNameTy s1) (FromName s1) uni fun ann
+             -> m (PLC.Program (FromNameTy s2) (FromName s2) uni fun ann)
+plcToOutName sng1 ((sng1 %~) -> Proved Refl) = pure
+plcToOutName SName SNamedDeBruijn = PLC.progTerm PLC.deBruijnTerm
+plcToOutName SNamedDeBruijn SName = PLC.runQuoteT . PLC.progTerm PLC.unDeBruijnTerm
+plcToOutName SDeBruijn SNamedDeBruijn = pure . PLC.programMapNames PLC.fakeTyNameDeBruijn PLC.fakeNameDeBruijn
+plcToOutName SNamedDeBruijn SDeBruijn = pure . PLC.programMapNames PLC.unNameTyDeBruijn PLC.unNameDeBruijn
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/Compile.hs:253]: These functions suck a bit. I need to rewrite them in another PR

File: plutus-core/executables/plutus/AnyProgram/Parse.hs (hunk 13 above)
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/Parse.hs]: Added reference, thanks

File: plutus-core/executables/plutus/AnyProgram/With.hs [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+instance Semigroup PLC.SrcSpan where
+-- FIXME: stub
+
+withMonoidA :: forall s r. SAnn s ->  ((Monoid (FromAnn s)) => r) -> r
+withMonoidA s r = case s of
+    SUnit -> r
+    SSrcSpan_ -> r
+
+instance Monoid PLC.SrcSpan where
+-- FIXME: stub
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/With.hs]: It's just that there is no sensible implementation of this ATM . We had a discussion on Slack about it some months ago

File: plutus-core/executables/plutus/Mode/PrintBuiltins.hs (hunk 14 above)
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/Mode/PrintBuiltins.hs:27]: @effectfully Maybe with your recent changes to use a `Map`  there, can we have sth like versions?

File: plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Core/Instance/Flat.hs [hunk 18]
```diff
@@ -247,6 +253,23 @@ sizeProgram (Program ann v t) sz = size ann $ size v $ sizeTerm t sz
...
+
+deriving newtype instance (Show name, GShow uni, Everywhere uni Show, Show fun, Show ann, Closed uni)
+    => Show (UnrestrictedProgram name uni fun ann)
+
+deriving via PrettyAny (UnrestrictedProgram name uni fun ann)
+    instance DefaultPrettyPlcStrategy (UnrestrictedProgram name uni fun ann) =>
+        PrettyBy PrettyConfigPlc (UnrestrictedProgram name uni fun ann)
+
+deriving newtype instance
+    (PrettyBy (PrettyConfigClassic PrettyConfigName) name, Everywhere uni PrettyConst, PrettyParens (SomeTypeIn uni), Closed uni, Pretty fun, Pretty ann)
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Core/Instance/Flat.hs:269]: I had problems with using type synonyms in those instances.

Especially in the head of the instance.

File: plutus-core/executables/plutus/Mode/PrintBuiltins.hs (hunk 8 above)
Comment (bezirg) [reply plutus-core/executables/plutus/Mode/PrintBuiltins.hs:31]: It is ported code, I didn't bother editing at all. It is a copy-paste

File: plutus-core/executables/plutus/Mode/PrintBuiltins.hs (hunk 9 above)
Comment (bezirg) [reply plutus-core/executables/plutus/Mode/PrintBuiltins.hs:22]: Not really, it was for you, so you don't have to review the following code :D

File: plutus-core/executables/plutus/GetOpt.hs [hunk 19]
```diff
@@ -0,0 +1,235 @@
...
+                    -> OptsFun
+overFileTypeDefault setter f arg = over lastFileType $ \ mFt ->
+    set (mapped . setter)
+    (f arg)
+    (mFt <|> Just defFileType)
+
+-- READING
+--------------------------------------------------
+
+-- FIXME: STUBS
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/GetOpt.hs]: Yes, I have to put because code will break

[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/GetOpt.hs]: I think I fixed it now!

File: plutus-core/executables/plutus/GetOpt.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/GetOpt.hs]: I left it in with a comment to about it for later

File: plutus-core/executables/plutus/AnyProgram/IO.hs [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+import Data.ByteString qualified as BS
+import Flat
+import System.IO
+import Data.Text.Encoding qualified as T
+import PlutusCore.Pretty qualified as PP
+import Prettyprinter
+import Prettyprinter.Render.Text
+import Control.Monad
+
+readProgram :: SLang s -> File s -> IO (FromLang s)
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/IO.hs]: Y the -x is supposed to be applied to everything after (including the output). If `-o` or `--stdout` is missing, then the last `-x` will be in effect

[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/IO.hs]: What is the expected type of this auctions?  Is it `Flat (Program ... DeBruijn .... ()`  ?

File: plutus-core/executables/plutus/AnyProgram/With.hs [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+    SName -> r
+    SDeBruijn -> r
+    SNamedDeBruijn -> r
+
+withFlatNT :: forall s r. SNaming s ->  (Flat (FromNameTy s) => r) -> r
+withFlatNT s r = case s of
+    SName -> r
+    SDeBruijn -> r
+    SNamedDeBruijn -> r
+
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/With.hs]: Y that is what @effectfully was suggesting. I managed to use his suggestion, but still I couldn't remove the boilerplate completely

File: plutus-core/executables/plutus/AnyProgram/Example.hs [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+                )]
+termExamples =
+    [ ("succInteger",  StdLib.succInteger)
+    , ("unitval",  StdLib.unitval)
+    , ("true",  StdLib.true)
+    , ("false",  StdLib.false)
+    , ("churchZero",  StdLib.churchZero)
+    , ("churchSucc",  StdLib.churchSucc)
+    ]
+
```
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/Example.hs:30]: I don't remember why I didn't port those as well....

File: plutus-core/executables/plutus/AnyProgram/IO.hs (hunk 20 above)
[the remark this answers is not in the document]
Comment (bezirg) [reply plutus-core/executables/plutus/AnyProgram/IO.hs]: I kind of hate annotations, mostly because of `Provenance`
